### PR TITLE
[Google Blockly] Show/hide with Blockly.Events.BlockDrag

### PR DESCRIPTION
**Pull request by ajpal**
Pure refactor with no user-visible changes

Instead of overriding and hooking into the `BlockDragger` directly, we can control the visibility of the trashcan by listening for [`Blockly.Events.BlockDrag`](https://developers.google.com/blockly/reference/js/Blockly.Events.BlockDrag?hl=en) events.

### CODE REVIEW

File: apps/src/blockly/addons/cdoBlockDragger.js [hunk 1]
```diff
@@ -1,55 +1,19 @@
...
-  /** Hide trashcan when drag ends.
-   * Also prevent blocks from ending with a negative position. Google allows the workspace to scroll in any direction,
-   * so negative block coordinates are valid, but we want to keep everything flowing down and to the right from (0,0).
+  /**
+   * Make sure disabled state is updated immediately.
    * @override
    */
   endDrag(e, currentDragDeltaXY) {
-    const wouldDeleteBlock = this.draggedConnectionManager_.wouldDeleteBlock();
-    // Don't let the block end with a negative position, unless it's getting deleted.
```
Comment (ajpal) [inline apps/src/blockly/addons/cdoBlockDragger.js:31]: No longer needed now that we have the ScrollOptions plugin

File: apps/src/blockly/addons/cdoTrashcan.js [hunk 2]
```diff
@@ -2,6 +2,52 @@ import GoogleBlockly from 'blockly/core';
...
+   */
+  init() {
+    super.init();
+    this.workspace_.addChangeListener(this.workspaceChangeHandler.bind(this));
+  }
+
+  workspaceChangeHandler(blocklyEvent) {
+    if (blocklyEvent.type === Blockly.Events.BLOCK_DRAG) {
+      let trashcanDisplay = 'none';
+      let toolboxVisibility = 'visible';
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoTrashcan.js:16]: Does `toolboxVisibility` determine if the toolbox panel is expanded? or is it used for something else?

File: apps/src/blockly/addons/cdoTrashcan.js [hunk 3]
```diff
@@ -2,6 +2,52 @@ import GoogleBlockly from 'blockly/core';
...
+      const isDraggingFromFlyout_ = !!Blockly.mainBlockSpace?.currentGesture_
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoTrashcan.js]: What does `flyout` mean here ?

File: apps/src/blockly/addons/cdoTrashcan.js [hunk 4]
```diff
@@ -2,6 +2,52 @@ import GoogleBlockly from 'blockly/core';
...
+        ?.flyout_;
+      if (!isDraggingFromFlyout_ && blocklyEvent.isStart) {
+        trashcanDisplay = 'block';
+        toolboxVisibility = 'hidden';
+      }
+
+      /**
+       * NodeList.forEach() is not supported on IE. Use Array.prototype.forEach.call() as a workaround.
+       * https://developer.mozilla.org/en-US/docs/Web/API/NodeList/forEach
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoTrashcan.js:27]: ah yes I've run into this issue with IE before. Thanks for adding the comment + link!

File: apps/src/blockly/addons/cdoTrashcan.js [hunk 5]
```diff
@@ -2,6 +2,52 @@ import GoogleBlockly from 'blockly/core';
...
+       */
+      Array.prototype.forEach.call(
+        // query selector for uncategorized toolbox contents
+        document.querySelectorAll('.blocklyFlyout .blocklyWorkspace'),
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoTrashcan.js:31]: Are the `uncategorized toolbox contents` in the toolbox panel itself?

File: apps/src/blockly/addons/cdoTrashcan.js [hunk 6]
```diff
@@ -2,6 +2,52 @@ import GoogleBlockly from 'blockly/core';
...
+      Array.prototype.forEach.call(
+        // query selector for categorized toolbox contents
+        document.querySelectorAll('.blocklyToolboxContents'),
+        function(x) {
+          x.style.visibility = toolboxVisibility;
+        }
+      );
+
+      document.querySelector('#trashcanHolder').style.display = trashcanDisplay;
+      const isDeletable = blocklyEvent.blocks.every(block =>
```
Comment (katleiahramos) [inline apps/src/blockly/addons/cdoTrashcan.js:45]: Does this mean that if one of the blocks that's being dragged in a group of blocks is not deletable, then the trashcan won't show / the group of blocks can't be deleted?

File: apps/src/blockly/addons/cdoTrashcan.js (hunk 5 above)
Comment (ajpal) [reply apps/src/blockly/addons/cdoTrashcan.js:31]: Yeah that's referring to the blocks in the toolbox.

image

File: apps/src/blockly/addons/cdoTrashcan.js (hunk 6 above)
Comment (ajpal) [reply apps/src/blockly/addons/cdoTrashcan.js:45]: Yeah, that group of blocks can't be deleted. The trashcan still shows, but there's an indicator to show that you're not allowed to delete:
Screen Shot 2021-12-01 at 2 02 25 PM

File: apps/src/blockly/addons/cdoTrashcan.js (hunk 2 above)
Comment (ajpal) [reply apps/src/blockly/addons/cdoTrashcan.js:16]: it indicates whether the toolbox contents will be hidden or visible. when a block is dragged, the trashcan shows up over the toolbox area and we hide the toolbox contents.

Comment (katleiahramos) [reply apps/src/blockly/addons/cdoTrashcan.js:16]: oh got it, so `trashcanDisplay` and `toolboxVisiblity` are kinda dependent on each other? would we ever have a scenario where we would want the toolbox to be visible and for the trashcan to show?

Comment (ajpal) [reply apps/src/blockly/addons/cdoTrashcan.js:16]: Right, yeah, they should always be opposite- when the trashcan is shown the toolbox contents are hidden, and vice versa. If they both showed at once it'd look like this, which would not be desirable:
image

Comment (katleiahramos) [reply apps/src/blockly/addons/cdoTrashcan.js:16]: ah I see! Thank you for breaking that down!!

**Comment by katleiahramos** [comment]
Pulled & Tested ✅  Thanks for answering all my questions! I definitely feel like I have a better understanding of this area of code and your changes make sense to me!

**Review by katleiahramos — APPROVED** [review]

**Pull request merged**